File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/GetAlbums.java
package expo.modules.medialibrary;

import android.content.Context;
import android.database.Cursor;
import android.os.AsyncTask;
import android.os.Bundle;
import android.provider.MediaStore;

import org.unimodules.core.Promise;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static expo.modules.medialibrary.MediaLibraryConstants.ERROR_UNABLE_TO_LOAD;
import static expo.modules.medialibrary.MediaLibraryConstants.ERROR_UNABLE_TO_LOAD_PERMISSION;
import static expo.modules.medialibrary.MediaLibraryConstants.EXTERNAL_CONTENT;

class GetAlbums extends AsyncTask<Void, Void, Void> {
  private final Context mContext;
  private final Promise mPromise;

  public GetAlbums(Context context, Promise promise) {
    mContext = context;
    mPromise = promise;
  }

  @Override
  protected Void doInBackground(Void... params) {

    final String[] projection = {MediaStore.Images.Media.BUCKET_ID, MediaStore.Images.Media.BUCKET_DISPLAY_NAME};
    final String selection = MediaStore.Files.FileColumns.MEDIA_TYPE + " != " + MediaStore.Files.FileColumns.MEDIA_TYPE_NONE;

    Map<String, Album> albums = new HashMap<>();

    try (Cursor asset = mContext.getContentResolver().query(
        EXTERNAL_CONTENT,
        projection,
        selection,
        null,
        MediaStore.Images.Media.BUCKET_DISPLAY_NAME)) {

      if (asset == null) {
        mPromise.reject(ERROR_UNABLE_TO_LOAD, "Could not get albums. Query returns null.");
      } else {
        final int bucketIdIndex = asset.getColumnIndex(MediaStore.Images.Media.BUCKET_ID);
        final int bucketDisplayNameIndex = asset.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME);

        while (asset.moveToNext()) {
          final String id = asset.getString(bucketIdIndex);
          if (!albums.containsKey(id)) {
            final String title = asset.getString(bucketDisplayNameIndex);
            albums.put(id, new Album(id, title));
          } else {
            albums.get(id).incrementCount();
          }
        }

        List<Bundle> result = albums.values().stream().map(Album::toBundle).collect(Collectors.toList());
        mPromise.resolve(result);
      }
    } catch (SecurityException e) {
      mPromise.reject(ERROR_UNABLE_TO_LOAD_PERMISSION,
        "Could not get albums: need READ_EXTERNAL_STORAGE permission.", e);
    } catch (IllegalArgumentException e) {
      mPromise.reject(ERROR_UNABLE_TO_LOAD, "Could not get albums.", e);
    }
    return null;
  }

  private class Album {
    private String id;
    private String title;
    private int count;

    Album(String id, String title) {
      this.id = id;
      this.title = title;
      this.count = 1;
    }

    void incrementCount() {
      this.count++;
    }

    Bundle toBundle() {
      Bundle album = new Bundle();
      album.putString("id", this.id);
      album.putString("title", this.title);
      album.putParcelable("type", null);
      album.putInt("assetCount", this.count);
      return album;
    }
  }
}

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/GetAlbums.kt
package expo.modules.medialibrary

import android.content.Context
import android.database.Cursor.FIELD_TYPE_NULL
import android.os.AsyncTask
import android.os.Bundle
import android.provider.MediaStore
import android.provider.MediaStore.Images.Media
import org.unimodules.core.Promise

internal open class GetAlbums(private val mContext: Context, private val mPromise: Promise)
  : AsyncTask<Void?, Void?, Void?>() {

  override fun doInBackground(vararg params: Void?): Void? {
    val projection = arrayOf(Media.BUCKET_ID, Media.BUCKET_DISPLAY_NAME)
    val selection = "${MediaStore.Files.FileColumns.MEDIA_TYPE} != ${MediaStore.Files.FileColumns.MEDIA_TYPE_NONE}"

    val albums = HashMap<String, Album>()

    try {
      mContext.contentResolver
        .query(
          MediaLibraryConstants.EXTERNAL_CONTENT,
          projection,
          selection,
          null,
          Media.BUCKET_DISPLAY_NAME)
        .use { asset ->
          if (asset == null) {
            mPromise.reject(MediaLibraryConstants.ERROR_UNABLE_TO_LOAD,
              "Could not get albums. Query returns null.")
            return@use
          }
          val bucketIdIndex = asset.getColumnIndex(Media.BUCKET_ID)
          val bucketDisplayNameIndex = asset.getColumnIndex(Media.BUCKET_DISPLAY_NAME)

          while (asset.moveToNext()) {
            val id = asset.getString(bucketIdIndex)

            if (asset.getType(bucketDisplayNameIndex) == FIELD_TYPE_NULL) {
              continue
            }

            val album = albums[id] ?: Album(
                id = id,
                title = asset.getString(bucketDisplayNameIndex)
              ).also {
                albums[id] = it
              }

            album.count++
          }

          mPromise.resolve(albums.values.map { it.toBundle() })
        }
    } catch (e: SecurityException) {
      mPromise.reject(MediaLibraryConstants.ERROR_UNABLE_TO_LOAD_PERMISSION,
        "Could not get albums: need READ_EXTERNAL_STORAGE permission.", e)
    } catch (e: RuntimeException) {
      mPromise.reject(MediaLibraryConstants.ERROR_UNABLE_TO_LOAD, "Could not get albums.", e)
    }
    return null
  }

  private class Album(private val id: String, private val title: String, var count: Int = 0) {
    fun toBundle() = Bundle().apply {
      putString("id", id)
      putString("title", title)
      putParcelable("type", null)
      putInt("assetCount", count)
    }
  }
}
